mission/mission_planner.cpp

```diff
@@ -0,0 +1,900 @@
+
+    init_success &= message_handler_.add_cmd_callback(  MAV_CMD_NAV_TAKEOFF, // 22
+                                                        MAVLINK_BASE_STATION_ID,
+                                                        MAV_COMP_ID_ALL,
+                                                        MAV_COMP_ID_ALL, // 0
+                                                        &nav_takeoff_callback,
+                                                        this );
+
+    init_success &= message_handler_.add_cmd_callback(  MAV_CMD_OVERRIDE_GOTO, // 252
+                                                        MAVLINK_BASE_STATION_ID,
+                                                        MAV_COMP_ID_ALL,
+                                                        MAV_COMP_ID_ALL, // 0
+                                                        &override_goto_callback,
+                                                        this );
+
+    init_success &= message_handler_.add_cmd_callback(  MAV_CMD_MISSION_START, // 300
+                                                        MAVLINK_BASE_STATION_ID,
+                                                        MAV_COMP_ID_ALL,
+                                                        MAV_COMP_ID_ALL, // 190
+                                                        &mission_start_callback,
+                                                        this );
+
+    if(!init_success)
+    {
+        print_util_dbg_print("[MISSION_PLANNER] constructor: ERROR\r\n");
+    }
+
+    return init_success;
+}
+
+bool Mission_planner::update(Mission_planner* mission_planner)
+{
+    Mav_mode mode_local = mission_planner->state_.mav_mode();
+
+    // Reset to standby if disarmed
+    if (!mission_planner->state_.is_armed() && mission_planner->internal_state() != STANDBY)
+    {
+        mission_planner->inserted_waypoint_ = Waypoint();
+        mission_planner->switch_mission_handler(mission_planner->inserted_waypoint_);
+    }
+
+    switch (mission_planner->state_.mav_state_)
+    {
+    case MAV_STATE_STANDBY:
+        if (mission_planner->internal_state() != STANDBY)
+        {
+            mission_planner->inserted_waypoint_ = Waypoint();
+            // Switch, don't insert as the mission isn't paused
+            mission_planner->switch_mission_handler(mission_planner->inserted_waypoint_);
+        }
+        mission_planner->critical_behavior_ = CLIMB_TO_SAFE_ALT;
+        mission_planner->critical_next_state_ = false;
+        break;
+
+    case MAV_STATE_ACTIVE:
+        mission_planner->critical_behavior_ = CLIMB_TO_SAFE_ALT;
+        mission_planner->critical_next_state_ = false;
+
+        mission_planner->state_machine();
+        break;
+
+    case MAV_STATE_CRITICAL:
+        // In MAV_MODE_VELOCITY_CONTROL, MAV_MODE_POSITION_HOLD and MAV_MODE_GPS_NAVIGATION
+        if (mode_local.is_guided())
+        {
+            mission_planner->critical_handler();
+        }
```

Comment:
Is guided is set only during hold position and auto. If it's not guided we are in attitude or velocity which has the pilot controlling the drone.